src/Symfony/Component/Utf8/GenericStringInterface.php

```diff
@@ -0,0 +1,159 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Utf8;
+
+use Symfony\Component\Utf8\Exception\ExceptionInterface;
+
+/**
+ * Represents a generic string implementation.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugo.hamon@sensiolabs.com>
+ *
+ * @throws ExceptionInterface
+ * 
+ * @final
+ */
+interface GenericStringInterface extends \IteratorAggregate
+{
+    public function __toString();
+
+    /**
+     * @return int|null
+     */
+    public function indexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function indexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return static|null
+     */
+    public function substringOf(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function substringOfIgnoreCase(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function lastSubstringOf(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function lastSubstringOfIgnoreCase(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static
+     */
+    public function reverse();
+
+    public function length(): int;
+
+    public function width(bool $ignoreAnsiDecoration = true): int;
+
+    public function isEmpty(): bool;
+
+    /**
+     * Splits the string into smaller chunks separated by a delimiter.
+     *
+     * @return static[]
+     */
+    public function explode(string $delimiter, int $limit = null): array;
+
+    /**
+     * @return static
+     */
+    public function toLowerCase();
+
+    /**
+     * @return static
+     */
+    public function toUpperCase();
+
+    /**
+     * @return static
+     */
+    public function toUpperCaseFirst($allWords = false);
+
+    /**
+     * @return static
+     */
+    public function toFoldedCase(bool $full = true);
+
+    /**
+     * @return static
+     */
+    public function substr(int $start = 0, int $length = null);
+
+    /**
+     * @return static
+     */
+    public function trim(string $charsList = null);
+
+    /**
+     * @return static
+     */
+    public function trimLeft(string $charsList = null);
+
+    /**
+     * @return static
+     */
+    public function trimRight(string $charsList = null);
+
+    /**
+     * @return static
+     */
+    public function append(string $suffix);
+
+    /**
+     * @return static
+     */
+    public function prepend(string $prefix);
+
+    /**
+     * @param string|string[] $from
+     * @param string|string[] $to
+     *
+     * @return static
+     */
+    public function replace($from, $to, int &$count = null);
```

Comment:
What about having two "replace" functions, one that takes only strings for from/to, and another that takes only array of strings? Too bad we don't have polymorphism in php here :)
replace and replaceMap?